sycl/doc/extensions/DeviceGlobal/cl_intel_global_variable_pointers.asciidoc

```diff
@@ -0,0 +1,196 @@
+
+== Name Strings
+
+`cl_intel_global_variable_pointers`
+
+== Contact
+
+Ben Ashbaugh, Intel (ben 'dot' ashbaugh 'at' intel 'dot' com)
+
+== Contributors
+
+// spell-checker: disable
+* Ben Ashbaugh, Intel
+* Jaroslaw Chodor, Intel
+* Vyacheslav Zakharin, Intel
+// spell-checker: enable
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation. All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback.
+When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+If you are interested in using this feature in your software product, please let us know!
+
+== Version
+
+Built On: {docdate} +
+Revision: 0.9.0
+
+== Dependencies
+
+This extension is written against the OpenCL API Specification Version 3.0.6.
+
+This extension requires support for the `cl_intel_unified_shared_memory`
+extension.
+
+== Overview
+
+The extension allows programmers to obtain pointers to global variables (aka
+program scope variables) of OpenCL Programs and read/write such global variables
+using host APIs.
+
+== New API Functions
+
+[source]
+----
+cl_int clGetDeviceGlobalVariablePointerINTEL(
+    cl_device_id device,
+    cl_program program,
+    const char *global_variable_name,
+    size_t *global_variable_size_ret,
+    void **global_variable_pointer_ret );
+----
+
+== New API Enums
+
+None.
+
+== New OpenCL C Functions
+
+None.
+
+== Modifications to the OpenCL API Specification
+
+Modify Section 5.8.9, *Program Object Queries* ::
++
+--
+
+The function
+----
+cl_int clGetDeviceGlobalVariablePointerINTEL(
+    cl_device_id device,
+    cl_program program,
+    const char *global_variable_name,
+    size_t *global_variable_size_ret,
+    void **global_variable_pointer_ret );
+----
+is used to retrieve a pointer to a user-defined global variable
+_global_variable_name_ which exists in built _program_.  The interpretation of
+_global_variable_name_ depends on how the _program_ was created:
+
+* If the _program_ was created with *clCreateProgramWithIL* from SPIR-V, there
+  are two cases:
+
+  - If the SPIR-V module declares the *GlobalVariableDecorationsINTEL*
+    capability, the implementation looks first for an *OpVariable* that is
+    decorated with *HostAccessINTEL* where the _Name_ operand is the same as
+    _global_variable_name_.
+
+  - The implementation next looks for an *OpVariable* that is decorated with
+    *LinkageAttributes* where the _Linkage Type_ is *Export* and the _Name_
+    operand is the same as _global_variable_name_.
+
+* If the _program_ was created in any other way, the interpretation of
+  _global_variable_name_ is implementation-defined.
+
+The returned pointer is a Unified Shared Memory pointer to the requested global
```

Comment:
How does this work, when the `device_global` was instantiated with the `device_image_scope` property, where the underlying memory is a value type rather than a USM pointer?

Reply:
This wording comes from @bashbaug, but I think I can answer.

First of all, this has nothing to do with the `device_image_scope` property since that is a SYCL property and this document is an OpenCL extension spec.  Your question is still relevant, though.  The `clGetDeviceGlobalVariablePointerINTEL()` API takes the name of a module scope (global) variable and returns a pointer to that variable.  That variable is *not* a USM buffer, but the API still says that the returned pointer is a USM pointer.  I found this confusing at first too.

I think the rationale is that the pointer returned by this API can be used as though it was a pointer to USM device memory.  This does make sense since the variable is allocated in global memory that is private to the device.  For example, the pointer can be passed to `clEnqueueMemcpyINTEL()`, which expects a USM pointer.